### PR TITLE
fix: preserve span heatmap cell dimensions

### LINKED ISSUE #1084: Span profiles heatmap boxes render thinly
Not 100% clear yet why those boxes render with such a slim height like this, but we need to investigate it further.

Image

### PR DESCRIPTION
**Pull request by simonswine**
Fixes #1084.

I noticed a few edge cases in the heatmap representation. This PR fixes them and validates the behavior with production responses in E2E tests.

- Preserve non-zero Y-bucket origins.
- Collapse duplicate Y-bucket starts and sum their counts.
- Use a single X calibration column instead of filling every timestamp gap.

### CODE REVIEW

File: docker-compose.e2e.yaml [hunk 1]
```diff
@@ -16,7 +16,7 @@ services:
       GF_INSTALL_PLUGINS: grafana-llm-app
       OPENAI_API_KEY: $OPENAI_API_KEY
       OPENAI_ORGANIZATION_ID: $OPENAI_ORGANIZATION_ID
-      GF_FEATURE_TOGGLES_ENABLE: ${GF_FEATURE_TOGGLES_ENABLE:-metricsFromProfiles,flameGraphWithCallTree,localizationForPlugins}
+      GF_FEATURE_TOGGLES_ENABLE: ${GF_FEATURE_TOGGLES_ENABLE:-metricsFromProfiles,flameGraphWithCallTree,localizationForPlugins,profilesHeatmap,profilesExemplars}
```
Comment (hugohaggmark) [inline docker-compose.e2e.yaml]: nit: we just opened a [PR](https://github.com/grafana/profiles-drilldown/pull/1090) that removes profilesExemplars feature toggle

File: package.json [hunk 2]
```diff
@@ -14,7 +14,7 @@
     "e2e:ci:prepare": "docker build --progress=auto -t pyroscope-app-plugin-e2e:latest -f Dockerfile.plugin.e2e .",
     "e2e:ci:server:down": "docker compose down",
     "e2e:ci:server:up": "docker compose -f docker-compose.e2e.yaml up --build -d",
-    "e2e:ci:server:up:react19": "GRAFANA_VERSION=dev-preview-react19 GRAFANA_IMAGE=grafana-enterprise GF_FEATURE_TOGGLES_ENABLE=metricsFromProfiles,react19 docker compose -f docker-compose.e2e.yaml up --build -d",
+    "e2e:ci:server:up:react19": "GRAFANA_VERSION=dev-preview-react19 GRAFANA_IMAGE=grafana-enterprise GF_FEATURE_TOGGLES_ENABLE=metricsFromProfiles,profilesHeatmap,profilesExemplars,react19 docker compose -f docker-compose.e2e.yaml up --build -d",
```
Comment (hugohaggmark) [inline package.json]: nit: we just opened a [PR](https://github.com/grafana/profiles-drilldown/pull/1090) that removes profilesExemplars feature toggle

File: src/pages/ProfilesExplorerView/components/SceneExploreServiceHeatmap/infrastructure/buildHeatmapDataFrames.ts [hunk 3]
```diff
@@ -85,22 +81,28 @@ export function buildHeatmapDataFrame(
...
   }
 
-  return lowerBucketStarts;
+  const yMin = [...countsByBucketStart.keys()].sort((a, b) => a - b);
+  return {
+    ...slot,
+    yMin,
+    counts: yMin.map((bucketStart) => countsByBucketStart.get(bucketStart) ?? 0),
+  };
 }
```
Comment (hugohaggmark) [inline src/pages/ProfilesExplorerView/components/SceneExploreServiceHeatmap/infrastructure/buildHeatmapDataFrames.ts]: It would be really awesome if we could add unit tests for these smaller functions, in addition to testing the higher-level `buildHeatmapDataFrame` function.

I think that would make the changes much easier to review and understand, especially for someone coming in without much context.

**Review by hugohaggmark — COMMENTED** [review]
Awesome work on this! 🎉 I'll leave the final approval to someone on the team who has the right context, but everything looks good from my side. 👍

File: docker-compose.e2e.yaml (hunk 1 above)
Comment (simonswine) [reply docker-compose.e2e.yaml]: Done in https://github.com/grafana/profiles-drilldown/pull/1089/commits/b1c1d296918ca88014f2690510d2c0ec476a64a3

File: package.json (hunk 2 above)
Comment (simonswine) [reply package.json]: https://github.com/grafana/profiles-drilldown/pull/1089/commits/b1c1d296918ca88014f2690510d2c0ec476a64a3

File: src/pages/ProfilesExplorerView/components/SceneExploreServiceHeatmap/infrastructure/buildHeatmapDataFrames.ts (hunk 3 above)
Comment (simonswine) [reply src/pages/ProfilesExplorerView/components/SceneExploreServiceHeatmap/infrastructure/buildHeatmapDataFrames.ts]: I have added  coverage for those. I think my main problem is that the heatmap semantics from the Visualization are not speced anywhere (at least I know of), so this is mostly experimentation to make sure the heatmap is rendering as intended (linear buckets in Y-axis, step aliged X-axis)

Comment (hugohaggmark) [reply src/pages/ProfilesExplorerView/components/SceneExploreServiceHeatmap/infrastructure/buildHeatmapDataFrames.ts]: thank you for taking the time to add those tests 🙌

**Review by L2D2Grafana — APPROVED** [review]
LGTM.

It is a bit hard to click the ones on the bottom.
Image

**Comment by simonswine** [comment]
> It is a bit hard to click the ones on the bottom.

Maybe we can force the y-axis to start at 0. That should leave a better gap? Will create a follow up
